### ISSUE TITLE
[BUG] reduce_col produces slightly different results across runs with same input

### ISSUE THREAD

**Issue by tyeah** [ml-explore/mlx#878]
**Describe the bug**
When I do mx.mean with indices other than the last dim, on float tensors, I get slightly different results across runs with the same input, as shown in the example below. Is this expected?

I thought it could be the order of floating point math, but when I check reduce_col.metal, it seems to me that for any given output element, the math ops are sequential?

**To Reproduce**

Include code snippet
```python
import mlx.core as mx

x = mx.random.normal((1, 512, 512, 3))
y = mx.mean(x, (1, 2))
y1 = mx.mean(x, (1, 2))
print(mx.max(mx.abs(y1 - y)))
# prints something like array(8.73115e-10, dtype=float32)
```

**Expected behavior**
Identical result across runs, as in PyTorch.

**Desktop (please complete the following information):**
 - OS Version: [e.g. MacOS 14.1]
 - Version [e.g. 0.8.0]

**Additional context**
reduce_row is always consistent across runs. reduce_col on cpu is also consistent.

**Comment by azhucea** [comment]
With more experiments, I found reduce_row, when row dimension is large enough, is also inconsistent across runs. For example:
```python
x = mx.random.normal((32, 1048576))
y = mx.mean(x, 1)
y1 = mx.mean(x, 1)
print(mx.max(mx.abs(z - z1)))
# prints something around 1e-8
```

**Comment by awni** [comment]
That's not really a bug. Its expected that run to run won't be bitwise identical as the atomics (and possibly SIMD reductions 🤔 ) are not deterministic.

**Comment by tyeah** [comment]
I see, it's the atomics.. Thank you!

If I really want bit identical for reduce, do you have any suggestions? 

One way I can think of is, transpose the indices into inner dimension, and do `x.copy()` to get a contiguous array, and reduce on that. I added copy method to array binding, but it doesn't give bit accurate results either. I guess `copy` is not doing the "contiguous copy"?

To be accurate, what I did was
```python
import mlx.core as mx

# This is not bit accurate
x = mx.random.normal((1, 512 * 512, 128))
x = x.swapaxes(0, 1).copy()
y = mx.mean(x, 2)
y1 = mx.mean(x, 2)
print(mx.max(mx.abs(y1 - y)))

# this is bit
x = mx.random.normal((1, 128 , 512 * 512))
y = mx.mean(x, 2)
y1 = mx.mean(x, 2)
print(mx.max(mx.abs(y1 - y)))
```

**Comment by awni** [comment]
> If I really want bit identical for reduce, do you have any suggestions?

- Use the CPU reduce `mx.mean(x, 2, stream=mx.cpu)`
- In think currently `a.flatten().reshape(a.shape)` will give you a contiguous array. But that may change in the future so use with caution

>  I guess copy is not doing the "contiguous copy"?

Indeed it's a shallow copy in terms of the underlying memory. There is no way to force a true "deep copy". You can hack it a bit with tricks like the above but we are not planning to add something like it to the API anytime soon.

**Comment by awni** [comment]
> If I really want bit identical for reduce, do you have any suggestions?

Could you say why? Is it just for debugging? 

I ask because we haven't really considered this in the past.. but it might be worth having a fully deterministic mode.. albeit possibly slower.

**Comment by tyeah** [comment]
Thanks for the suggestions!
> Use the CPU reduce mx.mean(x, 2, stream=mx.cpu)
I tried this. It's slower than gpu but acceptable :)
> In think currently a.flatten().reshape(a.shape) will give you a contiguous array. But that may change in the future so use with caution
Yes! This makes it fully deterministic and almost equally fast. Thank you!

> Could you say why? Is it just for debugging?
Yeah, for debugging and faster iteration of different techniques. For example, if I want to compare qlora vs normal lora, if training results are fully deterministic, I can compare learning curve and generation results of the first, say, 100 iterations. But with randomness, it's hard to say if the difference is caused by quantization or randomness, so I will likely need to wait for more iterations to compare results.

**Comment by tyeah** [comment]
FYI I think tensorflow had [the same issue](https://github.com/tensorflow/tensorflow/issues/2732) in its early days and got fixed.

**Comment by awni** [comment]
I think every framework has this issue :)

**Comment by tyeah** [comment]
Yeah. And I'm quite happy with the workaround you suggested for now. Thank you for all your help!
